exploration/selection-matching-options.md

```diff
@@ -0,0 +1,512 @@
+|---|---|---|---|---|---|
+| MF1 Compat | ? | ? | - | + | some say F-M not compat. |
+| Devlopers/Translators can control | +++ | - | - | - | D/Ts have the ability to influence or override selection |
+| Developers/Translator must control | - | +++ | ++ | +++ | D/Ts are required to manage selection order |
+| Visual Inspection | +++ | + | + | + | It is possible to order any matrix canonically, enabling visual inspection |
+| Complex matching (varies by locale) | - | +++ | +++ | +++ | Matrix explosion may conflict with manual ordering in FM |
+| Complex matching (multi-value) | - | +++ | ++ | ++ | F-M stops on first match; B-M gives developer full control of matching |
+| Translation tool variant order | - | +++ | +++ | +++ | Translations tools are required to maintain the order and/or provide for reordering that itself is remembers (eg. in the TM) |
+| Partial leverage on added keys | - | +++ | +++ | +++ | Changes or additions to matrix only affect some entries |
+| Programmable selection order | + | + | + | + | Selector authors can provide options for tailoring matches |
+
+
+### Example
+
+To fascilitate discussion, I will use this _message_ as an example:
+
+```
+match :plural($days) :plural($items) :plural($coins)
+when 0   *   *   {This opportunity has ended}
+when one 0   one {Congratulations, you earned {$coins} coin}
+when one 0   *   {Congratulations, you earned {$coins} coins}
+when one one one {You have {$days} day to purchase {$items} item to earn {$coins} coin}
+when one one *   {You have {$days} day to purchase {$items} item to earn {$coins} coins}
+when one *   one {You have {$days} day to purchase {$items} items to earn {$coins} coin}
+when one *   *   {You have {$days} day to purchase {$items} items to earn {$coins} coins}
+when *   0   one {Congratulations, you earned {$coins} coin}
+when *   0   *   {Congratulations, you earned {$coins} coins}
+when *   one one {You have {$days} days to purchase {$items} item to earn {$coins} coin}
+when *   one *   {You have {$days} days to purchase {$items} item to earn {$coins} coins}
+when *   *   one {You have {$days} day to purchase {$items} items to earn {$coins} coin}
+when *   *   *   {You have {$days} days to purchase {$items} items to earn {$coins} coins}
+```
+
+## First-Match
+
+First-Match selection evaluates the list of _keys_ row-by-row and selects the first _variant_ that matches the list of _selectors_ as the _pattern_.
+
+In the example message, the _variants_ are in a canonical order, so first-match produces the same order as best-match does.
+
+**Pros**
++ Allows developers to control the order of selection.
++ Allows translators to tailor the order of selection.
++ Can visually inspect match order.
++ May be more efficient when perfoming match (??)
+
+**Cons**
+- Requires developers to specify _variants_ in the correct order.
+- Requires translators to tailor the order of _variants_ if this is different from the source.
```

Comment:
No, it refers to the need for translators to tailor the order of the variants in their language/locale where required.

For example, I could have this for English:

```
when 0 { }
when one { }
when 2 { }
when * { }
```

In Russian the "exploded" list might be:

```
when 0 { }
when one { }
when 2 { }
when few { }
when many { }
when * { }
```

... but Russian requires that the `2` key appear before the `few`/`many` **and** `one` keys (so that it is evaluated before one of those other keys "capture" the number `2`. (In `ru` the `one` plural rule fires for numbers like 21, 31, 41, etc.)

Reply:
Ah, ok. I have been presuming that this "explosion" is mostly handled by tooling/automation, which would know how to sort the variants.